### PR TITLE
feat(lib): make lib no_std

### LINKED ISSUE #3391: Support nostd
### Problem

Have you considered adding support for nostd? Or will similar PRs be allowed to be merged?

### Expected behavior

_No response_

### PR DESCRIPTION
**Pull request by cathaysia**
close #3391

### CODE REVIEW

File: lib/binding_rust/ffi.rs [hunk 1]
```diff
@@ -9,6 +9,7 @@ include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
 include!("./bindings.rs");
 
 #[cfg(any(unix, target_os = "wasi"))]
+#[cfg(feature = "std")]
```
Comment (ObserverOfTime) [inline lib/binding_rust/ffi.rs:12]: Should be merged with the above condition and also added to the windows version.

File: lib/binding_rust/lib.rs [hunk 2]
```diff
@@ -2025,22 +2026,25 @@ impl Query {
...
-                        let regex = &string_values[p[2].value_id as usize];
-                        text_predicates.push(TextPredicateCapture::MatchString(
-                            p[1].value_id,
-                            regex::bytes::Regex::new(regex).map_err(|_| {
-                                predicate_error(row, format!("Invalid regex '{regex}'"))
-                            })?,
-                            is_positive,
-                            match_all,
-                        ));
+                        #[cfg(feature = "std")]
```
Comment (ObserverOfTime) [inline lib/binding_rust/lib.rs:2029]: This should apply to the entire match case.

File: lib/binding_rust/lib.rs [hunk 3]
```diff
@@ -1,22 +1,32 @@
 #![doc = include_str!("./README.md")]
-
+#![cfg_attr(not(feature = "std"), no_std)]
 pub mod ffi;
 mod util;
 
+#[cfg(not(feature = "std"))]
+extern crate alloc;
```
Comment (ObserverOfTime) [inline lib/binding_rust/lib.rs:7]: Is this required?

Comment (cathaysia) [reply lib/binding_rust/lib.rs:7]: yes, alloc crate not included by default, we need include it by hand.

ref: https://docs.rust-embedded.org/book/collections/index.html#using-alloc

> [!NOTE]
> rust seems has no `#![feature(alloc)]` at this time.

**Comment by ObserverOfTime** [comment]
It seems that regex [does](https://docs.rs/regex/latest/regex/index.html#ecosystem-features) support `no_std`, albeit with some performance loss.

**Comment by cathaysia** [comment]
yes, I add regex to no_std. But as a trade-off, regex itself is removed from the workspace. Because `default-features = false ` need to be set in the workspace. This will cause changes to other crates

**Comment by ObserverOfTime** [comment]
There are more features that need to be toggled, and I also see [this](https://github.com/rust-lang/regex/blob/master/Cargo.toml#L40-L41) note now, so maybe regex should just be disabled in `no_std` after all.

**Comment by cathaysia** [comment]
that note is outdated. I test this pr for --target=thumbv7m-none-eabi, and this passed.

File: lib/Cargo.toml [hunk 4]
```diff
@@ -26,10 +26,12 @@ include = [
 ]
 
 [features]
+default = ["std"]
+std = ["regex/std"]
 wasm = ["wasmtime-c-api"]
 
 [dependencies]
-regex.workspace = true
+regex = { version = "1.10.4", default-features = false, features = ["unicode"] }
```
Comment (ObserverOfTime) [inline lib/Cargo.toml:34]:
```suggestion
regex = { version = "1.10.4", default-features = false, features = ["perf", "unicode", "regex-syntax/unicode"] }
```

File: lib/Cargo.toml [hunk 5]
```diff
@@ -26,10 +26,12 @@ include = [
 ]
 
 [features]
+default = ["std"]
+std = ["regex/std"]
```
Comment (ObserverOfTime) [inline lib/Cargo.toml:30]:
```suggestion
std = ["regex/std", "regex-syntax/std"]
```

**Comment by cathaysia** [comment]
this pr should ready be merged.

**Comment by cathaysia** [comment]
@amaanq anything block this?

**Pull request merged**